File: app/src/main/java/fi/crewradio/Packet.kt
package fi.crewradio

import java.nio.ByteBuffer

/**
 * Wire format, version 4. 18-byte header, big-endian integers, followed by the sealed payload:
 *
 *     'P' 'T' | version u8 = 4 | codec u8 | ttl u8 | hops u8 | senderId int32 | seq int32 | time uint32
 *     then: nonce (12) | ciphertext | tag (16)
 *
 * The payload is AES-256-GCM under the crew's channel key ([ChannelCrypto]) with the header as
 * associated data, except the ttl byte, which relays rewrite in place ([aadOf]). `hops` is the
 * sender's original hop budget and is authenticated: a relay never lets ttl exceed it, so a
 * captured packet with its ttl bumped travels no further than the sender allowed. `time` is
 * the sender's wall clock in whole seconds since the Unix epoch, authenticated too: a receiver
 * drops a packet more than [REPLAY_WINDOW_S] off its own clock ([isFresh]) before any cache is
 * touched, so a recording cannot be played back later than that, however long the sender has
 * been quiet or the receiver has been off the channel. Without the key a packet cannot be
 * read or forged; no other header field can be altered in flight.
 *
 * codec: 0 = one 20 ms frame of PCM16LE 16 kHz mono, 1 = one 20 ms Opus packet,
 *        2 = a [Hello] roster heartbeat (no audio).
 * ttl:   hops this packet may still travel. A relay forwards only while it is > 1 and decrements
 *        it first, and only while the packet is within the relay's own hop limit of its origin
 *        ([Ingress.relayTtl]), so a flood over a large mesh is bounded and no peer can buy itself
 *        extra hops.
 *
 * Compatibility: there is no legacy decoding. Version 3 had no timestamp (14-byte header); its
 * packets fail [parse] and are dropped, so a crew must run the same app version on every phone.
 */
object Packet {
    const val HEADER = 18
    const val VERSION = 4
    /** Largest packet a peer may send: a sealed PCM frame with header is 686 bytes, Opus far less. Anything bigger is dropped unread. */
    const val MAX_SIZE = 1024
    /** How far a packet's [Header.time] may be from the receiver's clock, either way, in seconds. */
    const val REPLAY_WINDOW_S = 60

    /** What the payload is. PCM and OPUS carry audio; HELLO is the roster heartbeat. */
    enum class Codec(val id: Int) {
        PCM(0), OPUS(1), HELLO(2);

        companion object {
            /** Codec for a wire id, or null for an id this build does not know. */
            fun fromId(id: Int): Codec? = entries.firstOrNull { it.id == id }
        }
    }

    /**
     * Parsed header fields; the payload starts at [HEADER] in the same array. [hops] is the
     * authenticated origin budget, [time] the sender's clock as an unsigned 32-bit second count.
     */
    class Header(val senderId: Int, val seq: Int, val codec: Codec, val ttl: Int, val hops: Int, val time: Long)

    /**
     * Builds a complete packet; ttl and hops are clamped to the byte range, hops defaults to the
     * ttl the sender stamps, [time] is seconds since the epoch (the low 32 bits go on the wire).
     */
    fun encode(senderId: Int, seq: Int, codec: Codec, ttl: Int, payload: ByteArray, hops: Int = ttl, time: Long): ByteArray {
        val bb = ByteBuffer.allocate(HEADER + payload.size)
        bb.put('P'.code.toByte()).put('T'.code.toByte())
        bb.put(VERSION.toByte()).put(codec.id.toByte()).put(ttl.coerceIn(0, 255).toByte()).put(hops.coerceIn(0, 255).toByte())
        bb.putInt(senderId).putInt(seq).putInt(time.toInt())
        bb.put(payload)
        return bb.array()
    }

    /** Returns null for anything that is not a well-formed v4 packet with a non-empty payload. */
    fun parse(p: ByteArray): Header? {
        if (p.size <= HEADER || p.size > MAX_SIZE || p[0] != 'P'.code.toByte() || p[1] != 'T'.code.toByte()) return null
        if (p[2].toInt() != VERSION) return null
        val codec = Codec.fromId(p[3].toInt() and 0xFF) ?: return null
        val ttl = p[4].toInt() and 0xFF
        val hops = p[5].toInt() and 0xFF
        val bb = ByteBuffer.wrap(p, 6, 12)
        return Header(bb.int, bb.int, codec, ttl, hops, bb.int.toLong() and 0xFFFF_FFFFL)
    }

    /**
     * True if [time] is within [REPLAY_WINDOW_S] of [nowS], both in seconds. The difference is
     * taken modulo 2^32, so the wire's unsigned counter wrapping (in 2106) is a small distance,
     * not a jump; either side may be ahead.
     */
    fun isFresh(time: Long, nowS: Long): Boolean {
        val d = time.toInt() - nowS.toInt()             // wraps with the 32-bit counter
        return d in -REPLAY_WINDOW_S..REPLAY_WINDOW_S
    }

    /** The header as authenticated: a copy of the first [HEADER] bytes with the ttl zeroed, since relays change it. */
    fun aadOf(p: ByteArray): ByteArray = p.copyOf(HEADER).also { it[4] = 0 }

    /** Rewrites the ttl byte of an already encoded packet in place, clamped to the byte range. */
    fun setTtl(p: ByteArray, ttl: Int) {
        p[4] = ttl.coerceIn(0, 255).toByte()
    }
}

File: sk-plugin/lib/packet.js
// SPDX-License-Identifier: EUPL-1.2
"use strict";

/**
 * Crew Radio wire format, version 4, exactly as the app's Packet.kt and Hello.kt define it.
 *
 *   'P' 'T' | version u8 = 4 | codec u8 | ttl u8 | hops u8 | senderId int32 BE | seq int32 BE | time uint32 BE
 *   then: nonce (12) | ciphertext | tag (16)            (see crypto.js)
 *
 * The header is the AES-GCM associated data with the ttl byte zeroed, since relays rewrite it;
 * everything else, `time` included, is authenticated. `time` is the sender's wall clock in whole
 * seconds since the epoch: a receiver drops a packet more than REPLAY_WINDOW_S off its own clock
 * before any cache sees it, which is what keeps a recording from being played back later.
 * codec 0 = one 20 ms frame of PCM16LE 16 kHz mono, 1 = one Opus packet, 2 = a hello.
 */

const HEADER = 18;
const VERSION = 4;
const MAX_SIZE = 1024;
const REPLAY_WINDOW_S = 60;

const Codec = Object.freeze({ PCM: 0, OPUS: 1, HELLO: 2 });

/**
 * Builds the 18-byte header. `hops` is the sender's original budget and defaults to the ttl;
 * `time` is in seconds and defaults to the clock now (production stamps it explicitly).
 */
function encodeHeader({ senderId, seq, codec, ttl, hops = ttl, time = Math.floor(Date.now() / 1000) }) {
  const h = Buffer.alloc(HEADER);
  h[0] = 0x50; // 'P'
  h[1] = 0x54; // 'T'
  h[2] = VERSION;
  h[3] = codec & 0xff;
  h[4] = clampByte(ttl);
  h[5] = clampByte(hops);
  h.writeInt32BE(senderId | 0, 6);
  h.writeInt32BE(seq | 0, 10);
  h.writeUInt32BE(time >>> 0, 14);
  return h;
}

/** Parses a packet's header; null for anything that is not a well-formed v4 packet with a payload. */
function parseHeader(p) {
  if (!Buffer.isBuffer(p) || p.length <= HEADER || p.length > MAX_SIZE) return null;
  if (p[0] !== 0x50 || p[1] !== 0x54 || p[2] !== VERSION) return null;
  const codec = p[3];
  if (codec !== Codec.PCM && codec !== Codec.OPUS && codec !== Codec.HELLO) return null;
  return { codec, ttl: p[4], hops: p[5], senderId: p.readInt32BE(6), seq: p.readInt32BE(10), time: p.readUInt32BE(14) };
}

/** The header as authenticated: the first 18 bytes with the ttl zeroed. */
function aadOf(p) {
  const a = Buffer.from(p.subarray(0, HEADER));
  a[4] = 0;
  return a;
}

/**
 * True when a packet stamped `time` is within the replay window of the clock `nowS` (both in
 * whole seconds). The comparison is unsigned 32-bit and wrap-safe, like the app's.
 */
function isFresh(time, nowS) {
  const d = ((time >>> 0) - (nowS >>> 0)) | 0;   // signed distance, wraps with the counter
  return d >= -REPLAY_WINDOW_S && d <= REPLAY_WINDOW_S;
}

function clampByte(n) {
  return Math.max(0, Math.min(255, n | 0));
}

// ---- Hello v2: ver u8 = 2 | transports u8 | ttl u8 | versionCode uint16 BE | nameLen u8 | name UTF-8 (max 32 bytes) ----

const HELLO_VERSION = 2;
const HELLO_HEAD = 6;
const HELLO_MAX_NAME = 32;
const Transports = Object.freeze({ LAN: 1, BT: 2, AWARE: 4 });

/**
 * Encodes a hello; the name is cut to 32 UTF-8 bytes on a character boundary, as the app does.
 * `versionCode` is the sender's build; 0 means not applicable, which is what this plugin sends.
 */
function encodeHello({ name, transports, ttl, versionCode = 0 }) {
  const bytes = utf8Prefix(name, HELLO_MAX_NAME);
  const h = Buffer.alloc(HELLO_HEAD + bytes.length);
  h[0] = HELLO_VERSION;
  h[1] = transports & 0xff;
  h[2] = clampByte(ttl);
  h.writeUInt16BE(Math.max(0, Math.min(0xffff, versionCode | 0)), 3);
  h[5] = bytes.length;
  bytes.copy(h, HELLO_HEAD);
  return h;
}

/**
 * A name as the roster shows it, the same rule as the app's Hello: ISO control characters and
 * Unicode format characters (bidi overrides, zero-width joiners and the like, which can make a
 * name read as another) are dropped, runs of whitespace become one space, and the ends are trimmed.
 */
function sanitiseName(s) {
  return String(s ?? "").replace(/[\p{Cc}\p{Cf}]/gu, "").replace(/\s+/g, " ").trim();
}

/** Decodes a hello payload; null for anything off the contract. */
function decodeHello(p) {
  if (!Buffer.isBuffer(p) || p.length < HELLO_HEAD || p[0] !== HELLO_VERSION) return null;
  const nameLen = p[5];
  if (nameLen > HELLO_MAX_NAME || p.length !== HELLO_HEAD + nameLen) return null;
  let name;
  try {
    name = new TextDecoder("utf-8", { fatal: true }).decode(p.subarray(HELLO_HEAD));
  } catch {
    return null;
  }
  return { name: sanitiseName(name), transports: p[1], ttl: p[2], versionCode: p.readUInt16BE(3) };
}

function utf8Prefix(s, max) {
  let str = String(s);
  let b = Buffer.from(str, "utf8");
  while (b.length > max) {
    str = str.slice(0, -1);
    b = Buffer.from(str, "utf8");
  }
  return b;
}

module.exports = {
  HEADER, VERSION, MAX_SIZE, REPLAY_WINDOW_S, Codec, Transports, HELLO_MAX_NAME, HELLO_VERSION,
  encodeHeader, parseHeader, aadOf, isFresh, encodeHello, decodeHello, sanitiseName,
};
